controllers/lease_controller.go

```diff
@@ -0,0 +1,544 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	controllersconfig "github.com/gardener/etcd-druid/controllers/config"
+	"github.com/gardener/etcd-druid/pkg/common"
+	"github.com/gardener/etcd-druid/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// LeaseController reconciles compaction job
+type LeaseController struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	logger      logr.Logger
+	ImageVector imagevector.ImageVector
+	config      controllersconfig.CompactionConfig
+}
+
+// NewLeaseController creates a new LeaseController object
+func NewLeaseController(mgr manager.Manager, config controllersconfig.CompactionConfig) *LeaseController {
+	return &LeaseController{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		logger: log.Log.WithName("lease-controller"),
+		config: config,
+	}
+}
+
+// NewLeaseControllerWithImageVector creates a new LeaseController object
+func NewLeaseControllerWithImageVector(mgr manager.Manager, config controllersconfig.CompactionConfig) (*LeaseController, error) {
+	lc := NewLeaseController(mgr, config)
+	return lc.InitializeControllerWithImageVector()
+}
+
+// InitializeControllerWithImageVector will use LeaseController client to initialize image vector for etcd
+// and backup restore images.
+func (lc *LeaseController) InitializeControllerWithImageVector() (*LeaseController, error) {
+	imageVector, err := imagevector.ReadGlobalImageVectorWithEnvOverride(getImageYAMLPath())
+	if err != nil {
+		return nil, err
+	}
+	lc.ImageVector = imageVector
+	return lc, nil
+}
+
+// +kubebuilder:rbac:groups=druid.gardener.cloud,resources=etcds,verbs=get;list;watch
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;create;list;watch;update;patch;delete
+
+// Reconcile reconciles the compaction job.
+func (lc *LeaseController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I find it a bit difficult to read and understand this controller, e.g. whether cases such as reconciling a creation or update and reconciling a deletion of an object are handled appropriately, whether the status is always correctly maintained, etc. With its current "monolithic" implementation, it would also be rather difficult to cover this implementation with unit tests, rather than with envtest, and I understand this is something we might want to do in the future.

I would therefore like to make a general suggestion. I think it's very useful to structure controllers in terms of "reconciler" and "actuator". This is an approach that is already used in g/g and extensions because of its "separation of concerns" and testability advantages.
* The "reconciler" is responsible for reading the object that is being reconciled, managing finalizers, annotations, and status updates, and making the distinction between reconciling a creation / update (what we call "reconcile") and reconciling a deletion (what we call "delete"). Reconciliation logic is usually similar / generic and so often the same reconciler can be used with different actuators - but even if this is not the case, the distinction is still useful.
* The "actuator" is responsible for the actual reconciliation / deletion logic and returning an appropriately updated status (but not actually updating it).

I used this approach in the `EtcdCopyBackupsTaskController` though I stopped short of actually extracting an actuator, so the actuator is "hidden" in the `doReconcile` and `doDelete` methods. It would be however easy to extract these methods to a separate type implementing an `Actuator` interface, and then test both the (much simpler now) `Reconciler` and `Actuator` in isolation from each other using unit tests with generated mocks.

In addition, whenever an actuator needs to produce a set of values for a chart, it could make use of a `ValuesProvider` interface and implementation that can again be further tested in isolation, since the logic for producing the correct values is often one of the most important pieces. Other helper types can be introduced as well when it makes sense for similar reasons.

As further examples of this approach, please see our provider extensions as well as the `ManagedSeed` and `ManagedSeedSet` controllers in gardener:
* https://github.com/gardener/gardener/tree/master/pkg/gardenlet/controller/managedseed
* https://github.com/gardener/gardener/tree/master/pkg/controllermanager/controller/managedseedset

Reply:
Hello @stoyanr,
Thanks for providing details on how g/g is structuring their code.

IMHO there is no clear distinction between actuator and reconciler in the example that you have provided (e.g ManagedSeed). There are recent PRs which also move away from actuator for `BackupEntry`, `BackupBucket` etc. (https://github.com/gardener/gardener/pull/6926) 

Actuator is useful when you have common code in the reconciler that one wishes to re-use and then delegate the responsibility to reconcile different resources to resource specific actuators. IMO it should not be generally applied. The point of testability is important and for that writing smaller functions with one clear purpose and single responsibility are usually sufficient unless there is a violation of DRY where additional interfaces like `Actuator` can be thought of. Having said that we will keep your suggestions in mind.